### PR TITLE
Output structured comments for build events

### PR DESCRIPTION
**Pull request by pietroalbini**
This PR changes homu to output a JSON version of each build-related comment (build started, build failed, build completed) so bots can easily parse it. This is useful for example to automatically detect when a try build is completed, or to capture its hash automatically.

The JSON output is appended to each comment inside an HTML comment, like this:

```
<!-- homu: {"type": "MessageType", ...} -->
```

For example:

```
<!-- homu: {"type":"TryBuildCompleted","merge_sha":"2b723429a9fa15e4e1098567c68b754aa6ef6bc8","builders":{"checks-travis":"https://travis-ci.com/pietroalbini/test-repo/builds/100369263"}} -->
```

cc @GuillaumeGomez

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Awesome, thanks a lot!

File: homu/main.py [hunk 1]
```diff
@@ -177,8 +178,12 @@ def get_issue(self):
             issue = self.issue = self.get_repo().issue(self.num)
         return issue
 
-    def add_comment(self, text):
-        self.get_issue().create_comment(text)
+    def add_comment(self, comment):
+        if issubclass(comment.__class__, comments.Comment):
```
Comment (kennytm) [inline homu/main.py]:
```suggestion
        if isinstance(comment, comments.Comment):
```

File: homu/server.py [hunk 2]
```diff
@@ -558,19 +559,20 @@ def report_build_res(succ, url, builder, state, logger, repo_cfg):
...
-                           'Approved by: {}\nPushing {} to {}...'
-                           ).format(state.approved_by, state.merge_sha,
-                                    state.base_ref)
-                state.add_comment(comment)
+                state.add_comment(comments.BuildCompleted(
+                    approved_by=state.approved_by,
+                    base_ref=state.base_ref,
+                    builders=dict(
+                        (k, v["url"]) for k, v in state.build_res.items()
+                    ),
```
Comment (kennytm) [inline homu/server.py]: I'm quite sure we have a sufficiently new Python that supports dictionary comprehension...

```python
{k: v['url'] for k, v in build_res.items()}
```

**Comment by Mark-Simulacrum** [comment]
Looks good otherwise!

File: homu/comments.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        )
+
+
+class TryBuildFailed(Comment):
+    params = ["builder_url", "builder_name"]
+
+    def render(self):
+        return ":broken_heart: Test failed - [%s](%s)" % (
+            self.builder_name, self.builder_url
+        )
```
[the remark this answers is not in the document]
Comment (pietroalbini) [reply homu/comments.py]: Added the timeout comment.

**Comment by pietroalbini** [comment]
Addressed all review comments.

File: homu/server.py [hunk 4]
```diff
@@ -598,24 +598,34 @@ def report_build_res(succ, url, builder, state, logger, repo_cfg):
...
                     state.add_comment(':eyes: ' + desc)
             else:
-                comment = (test_comment + '\n' +
-                           'State: approved={} try={}'
-                           ).format(state.approved_by, state.try_)
-                state.add_comment(comment)
+                state.add_comment(comments.TryBuildCompleted(
+                    builders=dict(
+                        (k, v["url"]) for k, v in state.build_res.items()
+                    ),
```
Comment (kennytm) [inline homu/server.py]: This `dict` too

**Pull request merged**